scripts/cfos/validate.py

```diff
@@ -5,70 +5,188 @@
-                                                Column("MMSE_Biopsy_value", [IsDtypeValidation(int)]),
-                                                Column("MMSE_Biopsy_atDonorAge", [IsDtypeValidation(int)]),
-                                                Column("MMSE_Final_value", [IsDtypeValidation(int)]),
-                                                Column("MMSE_Final_atDonorAge", [IsDtypeValidation(int)]),
-                                                Column("APOE_value", [IsDtypeValidation(int)]),
-                                                Column("APOE_atDonorAge", [IsDtypeValidation(int)]),
-                                                Column("DataModality", [InListValidation(data_modality_options)]),
-                                                # file path columns must start with "gs://"
-                                                # TODO: check starts with not just for pattern
-                                                Column("summary_file_path", [MatchesPatternValidation("gs://")]),
-                                                Column("features_file_path", [MatchesPatternValidation("gs://")]),
-                                                Column("matrix_file_path", [MatchesPatternValidation("gs://")]),
-                                                Column("barcode_file_path", [MatchesPatternValidation("gs://")]),
-                                                Column("biosample_id_of_file", null_validation),
-                                                Column("library_id", null_validation),
-                                                Column("UMI_threshold", [IsDtypeValidation(int)]),
-                                                Column("follow_up_years", [IsDtypeValidation(int)]),
-                                                Column("has_disease", [InListValidation(has_disease_options)])
-                                                   #    Column("comment", [None])
-                                            ])
+#null_validation = CustomElementValidation(lambda d: d is not np.nan, 'this field cannot be null')
+null_validation = CustomSeriesValidation(lambda d: d.notna(), 'this field cannot be null')
 
 
+# Keys
+OPTIONAL_FIELD_KEY = "optional"
+VALUE_REQUIRED_KEY = "value_required"
+INT_ONLY_VAL_KEY = "integer_only"
+FIELD_TYPE_KEY = "field_type"
+PATTERN_TO_MATCH_KEY = "pattern_to_match"
+ALLOWED_VALUES_KEY = "allowed_values"
+IS_UNIQUE_KEY = "is_unique"
 
+# Values
+FILE_PATH_FIELD_TYPE_VAL = "file_path"
+NUMERIC_FIELD_TYPE_VAL = "number"
 
 
+# Generic functions used to check attributes
+def field_attribute_value_exists(field_id, field_dict, attribute_to_check):
+   if attribute_to_check not in field_dict[field_id]:
+      return False
 
+   attribute_value = str(field_dict[field_id][attribute_to_check]).strip()
+   return attribute_value is not None and attribute_value != ""
 
 
+def check_field_type(field_id, field_dict, expected_type):
+   type_key = FIELD_TYPE_KEY
```

Comment:
rather than duplicating this variable, you can modify the following line to be `if FIELD_TYPE_KEY not in field_dict[field_id]:`

and similarly row 40 to `field_type = field_dict[field_id][FIELD_TYPE_KEY]`